src/Sparrow/Memory.cs

```diff
@@ -1,183 +1,793 @@
+            0xFF, 0xFF, 0xFF, 0xFF,
+            0xFF, 0xFF, 0xFF, 0xFF,
+            0xFF, 0xFF, 0xFF, 0xFF,
+            0xFF, 0xFF, 0xFF, 0xFF,
+            0xFF, 0xFF, 0xFF, 0xFF,
+            0xFF, 0xFF, 0xFF, 0xFF,
+            0x00, 0x00, 0x00, 0x00,
+            0x00, 0x00, 0x00, 0x00,
+            0x00, 0x00, 0x00, 0x00,
+            0x00, 0x00, 0x00, 0x00,
+            0x00, 0x00, 0x00, 0x00,
+            0x00, 0x00, 0x00, 0x00,
+            0x00, 0x00, 0x00, 0x00,
+            0x00, 0x00, 0x00, 0x00
+        };
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static int CompareInline(void* p1, void* p2, int size)
+        internal static int CompareAvx2(void* p1, void* p2, int size)
         {
-            // If we use an unmanaged bulk version with an inline compare the caller site does not get optimized properly.
-            // If you know you will be comparing big memory chunks do not use the inline version. 
-            if (size > CompareInlineVsCallThreshold)
-                goto UnmanagedCompare;
+            // PERF: Given all the preparation that must happen before even accessing the pointers, even if we increase
+            // the size of the method by 10+ bytes, by the time we access the data it is already there in L1 cache.
+            Sse.Prefetch0(p1);
+            Sse.Prefetch0(p2);
 
+            // PERF: This allows us to do pointer arithmetic and use relative addressing using the 
+            //       hardware instructions without needed an extra register.
             byte* bpx = (byte*)p1;
-            byte* bpy = (byte*)p2;
+            nuint offset = (nuint)((byte*)p2 - bpx);
 
-            // PERF: This allows us to do pointer arithmetics and use relative addressing using the 
-            //       hardware instructions without needed an extra register.            
-            long offset = bpy - bpx;
+            nuint length = (nuint)size;
+            byte* bpxEnd = bpx + length;
+
+            uint matches;
 
-            if ((size & 7) == 0)
+            // PERF: The alignment unit will be decided in terms of the total size, because we can use the exact same code
+            // for a length smaller than a vector or to force alignment to a certain memory boundary. This will cause some
+            // multi-modal behavior to appear (specially close to the vector size) because we will become dependent on
+            // the input. The biggest gains will be seen when the compares are several times bigger than the vector size,
+            // where the aligned memory access (no penalty) will dominate the runtime. So this formula will calculate how
+            // many bytes are required to get to an aligned pointer.
+            nuint alignmentUnit = length >= (nuint)Vector256<byte>.Count ? (nuint)(Vector256<byte>.Count - (long)bpx % Vector256<byte>.Count) : length;
+            if ((alignmentUnit & (nuint)(Vector256<byte>.Count - 1)) == 0 || length is >= 32 and <= 512)
```

Comment:
`Vector256<byte>.Count` is appearing multiple times here. Can we introduce a variable for it?